### PR TITLE
NPSP: Latest Start Date field on Address updating to Today on simple address change

### PR DESCRIPTION
**Pull request by dfuller114**
# Critical Changes

# Changes

# Issues Closed

# Community Ideas Delivered

# Features Intended for Future Release

# Features for Elevate Customers

# New Metadata

# Deleted Metadata

### CODE REVIEW

**Comment by andrewyu-salesforce** [comment]
[W-10703257](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE00000ogdPxYAI/view)

File: force-app/main/domain/Households.cls [hunk 1]
```diff
@@ -137,12 +137,8 @@ public inherited sharing class Households extends fflib_SObjects2 {
         // the Account to include mapping of the Primary Address Type.
         //Deactivate the Trigger that syncs Addresses to Accounts anc Contacts,
         // since the Household Account and Contact are already configured correctly.
-        TDTM_ProcessControl.toggleTriggerState('Address__c', 'ADDR_Addresses_TDTM',
```
Comment (lparrott) [inline force-app/main/domain/Households.cls:140]: Removing these toggles is great! Let's update/remove the comment above to reflect the change.

File: force-app/main/domain/Addresses.cls [hunk 2]
```diff
@@ -925,9 +926,9 @@ public inherited sharing class Addresses extends fflib_SObjects2 {
     private Addresses addressesWithDefaultChanges() {
         List<Address__c> addressesWithDefaultChanges = new List<Address__c>();
         for (Address__c address : (List<Address__c>) getRecords()) {
-            NPSP_Address npspAddress = new NPSP_Address(address);
+            NPSP_Address npspAddress = new NPSP_Address(address, oldVersionOf(address));
 
-            if (npspAddress.isDefault()) {
+            if (npspAddress.hasDefaultChanged()) {
```
Comment (lparrott) [inline force-app/main/domain/Addresses.cls:931]: Great find and fix, we were ignoring "former" defaults before.

File: force-app/test/ADDR_Addresses_TEST.cls [hunk 3]
```diff
@@ -2073,8 +2073,7 @@ public with sharing class ADDR_Addresses_TEST {
     **********************************************************************************************************/
     @IsTest
     private static void addressShouldBePropagatedToAccountsAndContactsOnMultipleTriggerCalls() {
-        // Arrange
-        Integer totalHH = 210;
```
Comment (lparrott) [inline force-app/test/ADDR_Addresses_TEST.cls:2077]: Are we losing a test scenario here (multiple trigger calls), or is this no longer needed now that we removed the TDTM toggles?

File: force-app/test/ADDR_Addresses_TEST.cls [hunk 4]
```diff
@@ -2111,6 +2110,188 @@ public with sharing class ADDR_Addresses_TEST {
...
+                                        LIMIT 1];
+        Address__c previousDefaultAddress = [SELECT Default_Address__c, Latest_Start_Date__c, Latest_End_Date__c
+                                             FROM Address__c
+                                             WHERE Id = :defaultAddressId
+                                             LIMIT 1];
+        System.assertEquals(false, previousDefaultAddress.Default_Address__c, 'The previous default address should be' +
+                ' set to false.');
+        System.assertEquals(true, newDefaultAddress.Default_Address__c, 'The new default address should be' +
+                ' set to true.');
+        System.assertEquals(System.today(), previousDefaultAddress.Latest_End_Date__c, 'The latest end date on the ' +
```
Comment (lparrott) [inline force-app/test/ADDR_Addresses_TEST.cls:2177]: I'm getting picky, but can we also assert the Latest Start Date for the former default? It should not be changed in this case.

**Review by lparrott — COMMENTED** [review]
Looks great! I left some minor suggestions, no real blockers.

File: force-app/main/domain/Households.cls (hunk 1 above)
Comment (andrewyu-salesforce) [reply force-app/main/domain/Households.cls:140]: Let's make sure the trigger toggle removal will not introduce some unnecessary SOQL/DML update. We had the issue inserting/deleting blank address.

File: force-app/test/ADDR_Addresses_TEST.cls [hunk 5]
```diff
@@ -2111,6 +2110,188 @@ public with sharing class ADDR_Addresses_TEST {
...
+        System.assertEquals(true, testContact.Current_Address__r.Default_Address__c, 'The address should be the default address of ' +
+                'the household.');
+
+        Address__c defaultAddress = [SELECT Default_Address__c
+                                     FROM Address__c
+                                     WHERE Id = :testContact.Current_Address__c];
+
+        Test.startTest();
+            defaultAddress.Default_Address__c = false;
+            update defaultAddress;
```
Comment (andrewyu-salesforce) [inline force-app/test/ADDR_Addresses_TEST.cls:2207]: Should remove the indentation here

File: force-app/test/ADDR_Addresses_TEST.cls [hunk 6]
```diff
@@ -2111,6 +2110,188 @@ public with sharing class ADDR_Addresses_TEST {
...
+            insert testAddresses[1];
+        Test.stopTest();
+
+        Address__c updatedTestAddress = [SELECT Default_Address__c, Latest_Start_Date__c, Latest_End_Date__c
+                                            FROM Address__c
+                                            WHERE Id = :testAddresses[1].Id
+                                            LIMIT 1];
+        System.assertEquals(false, updatedTestAddress.Default_Address__c, 'The previous default address should be' +
+                ' set to false.');
+        System.assertEquals(null, updatedTestAddress.Latest_End_Date__c, 'The latest end date on the ' +
```
Comment (andrewyu-salesforce) [inline force-app/test/ADDR_Addresses_TEST.cls:2250]: Should remove the indentation here

File: force-app/test/ADDR_Addresses_TEST.cls [hunk 7]
```diff
@@ -2111,6 +2110,188 @@ public with sharing class ADDR_Addresses_TEST {
...
+
+        Address__c testAddress = [SELECT Latest_Start_Date__c, Latest_End_Date__c
+                                  FROM Address__c
+                                  WHERE Id = :testAddresses[1].Id
+                                  LIMIT 1];
+        Test.startTest();
+            testAddress.Latest_Start_Date__c = System.today() - 2;
+            testAddress.Latest_End_Date__c = System.today() - 1;
+
+            update testAddress;
```
Comment (andrewyu-salesforce) [inline force-app/test/ADDR_Addresses_TEST.cls:2280]: Should remove the indentation here

File: force-app/test/ADDR_Addresses_TEST.cls (hunk 3 above)
Comment (dfuller114) [reply force-app/test/ADDR_Addresses_TEST.cls:2077]: I actually think @lparrott has a good point looking back at this. The intent was to max the default collection of trigger context records before a second trigger call is required (200). I'll revert this.

File: force-app/test/ADDR_Addresses_TEST.cls [hunk 8]
```diff
@@ -2111,6 +2110,200 @@ public with sharing class ADDR_Addresses_TEST {
...
+
+        Test.stopTest();
+
+        Address__c updatedTestAddress = [SELECT Default_Address__c, Latest_Start_Date__c, Latest_End_Date__c
+                                         FROM Address__c
+                                         WHERE Id = :testAddresses[1].Id
+                                         LIMIT 1];
+        System.assertEquals(false, updatedTestAddress.Default_Address__c, 'The previous default address should be' +
+                ' set to false.');
+        System.assertEquals(System.today() - 1, updatedTestAddress.Latest_End_Date__c, 'The latest end date on the ' +
```
Comment (npsp-reedestockton) [inline force-app/test/ADDR_Addresses_TEST.cls:2301]: These two assert messages do not match the assertion.

Comment (dfuller114) [reply force-app/test/ADDR_Addresses_TEST.cls:2301]: good catch!

**Review by andrewyu-salesforce — APPROVED** [review]

**Review by npsp-reedestockton — APPROVED** [review]
Nice work, @dfuller114!

**Pull request merged**